accepted/2020/form-factors.md

```diff
@@ -0,0 +1,206 @@
+
+### IL Linking
+
+The [IL linker](https://github.com/mono/linker) can generate optimized applications that fail at runtime, due to not being able to reliably analyze various problematic code patterns (largely centered on reflection use). It is not possible or desirable to simply invest more to improve the linker, as a strategy. Instead, build-time source generators will be the recommended mitigation for arbitrary reflection use.
+
+The barrier for entry for building source generators is very high today. There is no standard solution and only a few libraries can afford to build a custom one. Razor pages, gRPC, WPF data binding, and Data-Contract serializers each have their own custom build-time source generator solutions today. First-class Roslyn support for [source generators](https://github.com/dotnet/roslyn/blob/master/docs/features/source-generators.md) will lower the barrier of entry for components that need to generate source code.
+
+The existing annotations that provide hints to the IL linker about what to keep or what is safe to link are going to stay, but their use is going to be less preferred as compared to source generators. They provide a less predictable experience that is hard for users to get right.
+
+We will introduce a linker compatibility analyzer to detect code patterns that cannot be reliably analyzed by the linker. This will still be needed for libraries that do not or cannot adopt the source generation technique.
+
+### Single File
+
+The single-file form factor bundles all code into a single binary that the code is directly executed from. Single-file apps are easier to deploy and distribute. Many users have asked for this form factor. We will focus on producing traditional immutable single-file executables, and provide guidelines for application developers to target to successfully target the single-file form factor.
+
+We will not attempt to re-create application virtualization solutions such as [BoxedApp](https://www.boxedapp.com/docs/index.html) or (`App-V`)[https://docs.microsoft.com/en-us/windows/application-management/app-v/appv-for-windows] that can make unaware applications deployable as single-file executables by virtualizing the operating system. Application virtualization solutions are a unique domain of sophisticated technology, outside the realm of .NET runtime.
+
+Single-file compatibility analyzers will detect the use of APIs that are known to cause single file deployments to break, such as `Assembly.LoadFile(path)` or `Assembly.Location`. New APIs, such as [API that returns path to the launching .exe path](https://github.com/dotnet/runtime/issues/30448), will be introduced to simplify mitigation of these issues.
+
+Managed C++ is incompatible with the single file form factor and addressing this incompatibility in the tooling is prohibitively expensive. Projects written in managed C++ (such as part of WPF) would have to be rewritten in C# or other managed language to be single-file compatible.
```

Comment:
@jkotas could you link to or expand on why this is the case?

Reply:
Managed C++ binaries contain both managed and native code together. We are not able to load them from memory like pure managed assemblies. They must be handled same way as native libraries and either be in its own file on disk or statically linked into the single .exe. Statically linking them into the single .exe is difficult. It would require special support in Visual Studio linker.

Reply:
@jkotas this implies all Managed C++ assemblies are mixed mode assemblies does it not? This was what threw me off, it would be clearer if this was specifically mentioned, if that is pure managed c++ assemblies are not an issue :)

Reply:
`-clr:pure` mode of managed C++ is not supported for .NET Core. This is documented at https://docs.microsoft.com/en-us/dotnet/core/porting/cpp-cli#ccli-net-core-limitations .

I agree with you that `-clr:pure` would work better for this. This comes back to costs `-clr:pure` is used very rarely and it is much more expensive for the C++ team to maintain than `-clr`. It is the reason behind this.

Reply:
@jkotas I have many questions 😅 I understand that pure was dropped, but you can still have pure managed c++ assemblies but built for Framework, which however still work in Core. I don't understand the exact requirements for that to be the case, but assumed this was the case for pure managed c++ assemblies, is that true? 

We have a lot of third-party dependencies (from hardware vendors) that are c++/cli only if these work on Core can we use Core. So just trying to wrap my head around the conditions and limitations here. 😊

Reply:
Yes, it will likely work. https://docs.microsoft.com/en-us/dotnet/core/porting/third-party-deps#net-framework-compatibility-mode talks about it.